ign-rviz-plugins/src/rviz/plugins/tf_display.cpp

```diff
@@ -0,0 +1,206 @@
+    std::bind(&TFDisplay::callback, this, std::placeholders::_1));
+}
+
+rendering::ArrowVisualPtr TFDisplay::createTfArrow()
+{
+  rendering::ArrowVisualPtr arrow = this->scene->CreateArrowVisual();
+  rendering::VisualPtr head = arrow->Head();
+  head->SetMaterial(this->scene->Material("Default/TransPink"));
+  rendering::VisualPtr shaft = arrow->Shaft();
+  shaft->Scale(0.5, 0.5, 1.5);
+  arrow->SetOrigin(0, 0, -0.75);
+  shaft->SetMaterial(this->scene->Material("Default/TransYellow"));
+  return arrow;
+}
+
+rendering::VisualPtr TFDisplay::createVisualFrame()
+{
+  rendering::VisualPtr visualFrame = this->scene->CreateVisual();
+
+  // Add axis
+  rendering::AxisVisualPtr axis = this->scene->CreateAxisVisual();
+  axis->SetLocalScale(0.5);
+  visualFrame->AddChild(axis);
+
+  // Add arrow
+  rendering::ArrowVisualPtr arrow = this->createTfArrow();
+  arrow->SetLocalPosition(0, 0, 0);
+  arrow->SetVisible(false);
+
+  visualFrame->AddChild(arrow);
+
+  // Add text
+  rendering::TextPtr frameName = this->scene->CreateText();
+  frameName->SetTextString("frame");
+  frameName->SetShowOnTop(true);
+  frameName->SetTextAlignment(
+    rendering::TextHorizontalAlign::CENTER,
+    rendering::TextVerticalAlign::CENTER);
+  frameName->SetCharHeight(0.15);
+  visualFrame->AddGeometry(frameName);
+
+  return visualFrame;
+}
+
+bool TFDisplay::eventFilter(QObject * object, QEvent * event)
+{
+  if (event->type() == gui::events::Render::kType) {
+    updateTF();
+  }
+
+  return QObject::eventFilter(object, event);
+}
+
+void TFDisplay::updateTF()
+{
+  std::lock_guard<std::mutex>(this->lock);
+  std::vector<std::string> frameIds;
+  frameManager->getFrames(frameIds);
+
+  for (int i = tfRootVisual->ChildCount(); i < static_cast<int>(frameIds.size()); ++i) {
+    rendering::VisualPtr visualFrame = this->createVisualFrame();
+    this->tfRootVisual->AddChild(visualFrame);
+  }
+
+  for (int i = 0; i < static_cast<int>(frameIds.size()); ++i) {
+    ignition::math::Pose3d pose, parentPose;
+
+    rendering::VisualPtr visualFrame = std::dynamic_pointer_cast<rendering::Visual>(
+      this->tfRootVisual->ChildByIndex(i));
+
+    bool result = this->frameManager->getFramePose(frameIds[i], pose);
+
+    // Set Frame Text
+    rendering::TextPtr frameName = std::dynamic_pointer_cast<rendering::Text>(
+      visualFrame->GeometryByIndex(0));
+    frameName->SetTextString(frameIds[i]);
+
+    visualFrame->SetLocalPosition(pose.Pos());
+
+    rendering::AxisVisualPtr axis = std::dynamic_pointer_cast<rendering::AxisVisual>(
+      visualFrame->ChildByIndex(1));
+    axis->SetLocalRotation(pose.Rot());
+
+    result = this->frameManager->getParentPose(frameIds[i], parentPose);
+    if (result) {
+      rendering::ArrowVisualPtr arrow = std::dynamic_pointer_cast<rendering::ArrowVisual>(
+        visualFrame->ChildByIndex(
+          0));
+      math::Vector3d dir = parentPose.Pos() - pose.Pos();
+      double dist = dir.Length();
+      if (dist > 0.25) {
```

Comment:
magic number ?

Reply:
Haha. I just a small value I picked up as the arrow is not visible generally when the distance is very less. Disabled arrow visual as it might affect the render cycle (maybe I am overthinking on this). Thoughts?

Reply:
can you include this as a constant?